File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.py
class Solution:
    def prims(self, adj, V):
        pq = [(0, 0)]
        vis = [False] * V
        sum_cost = 0
        
        while pq:
            w, u = heapq.heappop(pq)
            if vis[u]:
                continue
            vis[u] = True
            sum_cost += w
            for nei, nwt in adj[u]:
                if not vis[nei]:
                    heapq.heappush(pq, (nwt, nei))
        
        return sum_cost

    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        V = len(points)
        adj = [[] for _ in range(V)]
        for i in range(V):
            for j in range(i + 1, V):
                x1, y1 = points[i]
                x2, y2 = points[j]
                d = abs(x1 - x2) + abs(y1 - y2)
                adj[i].append((j, d))
                adj[j].append((i, d))
        return self.prims(adj, V)

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.java
class Solution {
    static class Pair {
        int weight, node;
        Pair(int weight, int node) {
            this.weight = weight;
            this.node = node;
        }
    }
    
    public int prims(List<List<Pair>> adj, int V) {
        PriorityQueue<Pair> pq = new PriorityQueue<>(Comparator.comparingInt(p -> p.weight));
        pq.offer(new Pair(0, 0));
        boolean[] vis = new boolean[V];
        int sum = 0;
        
        while (!pq.isEmpty()) {
            Pair p = pq.poll();
            int w = p.weight;
            int u = p.node;
            if (vis[u]) {
                continue;
            }
            vis[u] = true;
            sum += w;
            for (Pair v : adj.get(u)) {
                int nei = v.node;
                int nwt = v.weight;
                if (!vis[nei]) {
                    pq.offer(new Pair(nwt, nei));
                }
            }
        }
        return sum;
    }
    
    public int minCostConnectPoints(int[][] points) {
        int V = points.length;
        List<List<Pair>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }
        for (int i = 0; i < V; i++) {
            for (int j = i + 1; j < V; j++) {
                int x1 = points[i][0];
                int y1 = points[i][1];
                int x2 = points[j][0];
                int y2 = points[j][1];
                
                int d = Math.abs(x1 - x2) + Math.abs(y1 - y2);
                adj.get(i).add(new Pair(d, j));
                adj.get(j).add(new Pair(d, i));
            }
        }
        return prims(adj, V);
    }
}